benchmarks/distributed/intra_node_comm/allgather_matmul.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+# This benchmark contains an example for using IntraNodeComm to implement efficient
+# fused allgather_matmul (inspired by https://dl.acm.org/doi/pdf/10.1145/3567955.3567959 and
+# @lw's efficient GPU implementation in xformers). The purpose of the example is to drive the
+# development of relevant primitives and serve as a recipe for interested users.
+#
+# The benchmark can be executed as follows:
+#   RDZV_ID=$(uuidgen) torchrun --nproc-per-node 8 allgather_matmul.py
+import os
+
+import torch
+import torch._C._distributed_c10d as c10d
+
+M = 16384
+N = 8192
+K = 2752
+
+WARMUP_ITERS = 200
+BENCH_ITERS = 50
+
+
+comm = None
+internal_stream = None
+internal_event = None
+
+
+def allgather_matmul(A_slice, B, out, rank, world_size):
+    """
+    Equivalent to `torch.matmul(dist.all_gather(A_slice), B)`.
+    """
+    buf_0 = torch.empty_like(A_slice)
+    buf_1 = torch.empty_like(A_slice)
+    out_slices = [
+        out[i : i + A_slice.shape[0]]
+        for i in range(0, world_size * A_slice.shape[0], A_slice.shape[0])
+    ]
+
+    # Perform matmul with the local input slice
+    torch.matmul(A_slice, B, out=out_slices[rank])
+
+    # In another stream, copy the local input slice into the intra-node
+    # buffer. After the barrier, all peers' input slices are accessible
+    # via their intra-node buffer without requiring synchronization.
+    with torch.cuda.stream(internal_stream):
+        comm.put(A_slice)
```

Comment:
hmmm why we don't specify the offset when doing `put` or `get`? I think if we don't then all ranks would put its own slice to the beginning of the intra node buffer?

Reply:
> I think if we don't then all ranks would put its own slice to the beginning of the intra node buffer?

Yes this is intentional. What we are doing is essentially a "one-sided" broadcast where the sender puts data into its buffer and all receivers pull from it.

> hmmm why we don't specify the offset when doing put or get?

Imagine we do a scatter in a similar fashion (i.e. receiver pulled) where slices are produced in stages. At the beginning, we would put slice 0 at the beginning of the buffer. When slice 1 is ready, we could evict slice 0 and replace it with slice 1. However this requires additional synchronization with the consumer of slice 0. If the IPC buffer is abundant, instead, we could put slice 1 into the buffer at an offset.